support/scripts/mkefi.py

```diff
@@ -0,0 +1,267 @@
+           r'(' + HEXNUM_EXP + r')' + r'\s*' +
+           HEXNUM_EXP + r'\s*' +
+           r'(' + HEXNUM_EXP + r')' + r'\s*' +
+           r'(' + HEXNUM_EXP + r')' + r'\s*' +
+           r'(' + r'[RWE ]+' + r')' + r'\s*' +
+           HEXNUM_EXP + r'$')
+    out = subprocess.check_output(["readelf", "-l", elf],
+                                   stderr=subprocess.DEVNULL)
+    re_out = re.findall(exp, out.decode('ASCII'), re.MULTILINE)
+
+    return [{
+                "Offset"    : int(r[0], 16),
+                "VirtAddr"  : int(r[1], 16),
+                "FileSiz"   : int(r[2], 16),
+                "MemSiz"    : int(r[3], 16),
+                "Flags"     : r[4].replace(' ', '')
+            } for r in re_out]
+
+def main():
+    parser = argparse.ArgumentParser(
+    description='Update the fake PE32 header with the required metadata to'
+                'be bootable by a UEFI environment and overwrite the ELF64'
+                'Header with an empty MS-DOS stub.')
+    parser.add_argument('elf', help='path to ELF64 binary to process')
+    opt = parser.parse_args()
+
+    # We need to operate on the debug image for symbol values. But as far as
+    # ELF sections go, using the final image is enough.
+    elf_dbg = opt.elf + ".dbg"
+
+    # Fetch base address, end address and bss start address from ld script
+    base_addr = get_sym_val(elf_dbg, r'_base_addr')
+    bss_addr = get_sym_val(elf_dbg, r'__bss_start')
+
+    # Consider the first PT_LOAD Segment as the first in the file.
+    ld_phdrs = get_loadable_phdrs(opt.elf)
+
+    # Make sure they are sorted by their addresses
+    ld_phdrs = sorted(ld_phdrs, key=lambda x: x['VirtAddr'])
+    end_addr = ld_phdrs[-1]["VirtAddr"] + ld_phdrs[-1]["MemSiz"]
+
+    # Again, all addresses relative to the very base of the file, because PE
+    # loading considers the MZ/PE/COFF headers as the first thing loaded.
+    # Use the first function in our EFI stub as the entry point.
+    entry_addr = get_sym_val(elf_dbg, r'uk_efi_entry64')
+
+    # PE is loaded by sections, thus the fake PE Header encodes the PT_LOAD's
+    # as follows:
+    # - dummy .reloc section since UEFI wants relocatable PE's
+    # - all the other ELF Program Headers as PE sections with all permissions
+    # enabled (RWX)
+    with open(opt.elf, 'r+b') as f:
+        elf_file = f.read()
+
+        # Get Machine
+        if sum(elf_file[18:20]) == ELF_MACHINE["arm64"]:
+            MZ_PE_HDR["Machine"][0] = PE_MACHINE["arm64"]
+        else:
+            MZ_PE_HDR['Machine'][0] = PE_MACHINE["x86_64"]
+
+        # Also add the fake .reloc section
+        MZ_PE_HDR["NumberOfSections"][0] = len(ld_phdrs) + 1
+        MZ_PE_HDR["ImageBase"][0] = base_addr
+        MZ_PE_HDR["SizeOfHeaders"][0] = PE_SEC_SZ * (len(ld_phdrs) + 1)
+        MZ_PE_HDR["SizeOfHeaders"][0] += 334  # See MZ_PE_HDR Offset comments
+        MZ_PE_HDR["SizeOfHeaders"][0] = MZ_PE_HDR["SizeOfHeaders"][0] & ~0xfff
```

Comment:
From around this point on this gets a bit hard to review. I'm sure it's possible with a bit of effort, but let's say it's not something that you go through casually. A bit more commenting would definitely help.